feat: Add Schema Visualizer to Supabase Studio

kamilogorek opened the pull request:
## What kind of change does this PR introduce?

This PR introduces a new sub-page called `Schemas` to the `Database` page.
It visualizes the selected schema (`public` by default) as an interactive graph.

## What is the new behavior?

![image](https://github.com/supabase/supabase/assets/1523305/55ba50d5-8ca8-40df-8582-9d9cb76e6c98)

## Additional context

There are some improvements and features that can be added (and better styling!), but for the rest of the community to jump in, the main functionality has to be in place.

Would appreciate some other folks testing it out locally, as due to its generative nature, the graph might be rendered in a funky way in case someone has a very complex schema in place.

Please note that I'm gone until Monday 17th starting tomorrow, so won't be able to respond to the feedback or help with setting things up till then.

## Possible improvements

- [ ] better styling overall
- [x] adding indicators
  - [x] primary key (right now its just a colored border)
  - [x] foreign key
  - [x] is nullable
  - [x] is identity
  - [x] unique
- [x] show truncated column names on hover (right now it shows as a native title attribute but would be nice to make it instant as a "floating" label
- [ ] a more complex addition would be to render a foreign table as a full node (although not sure if it's a good idea tbh, maybe just upgrade that node to be shown as a 3-line node with schema/table/column rows + column type and same indicators as above
- [ ] adding <Controls /> component with matching styles - https://reactflow.dev/docs/api/plugin-components/controls/; I had it in the initial implementation but didn't find it that useful tbh, but maybe with just a custom "force layout" button and "center" it might be actually helpful, it'd have to be a design UX decision

Closes https://github.com/supabase/supabase/issues/15585

kamilogorek committed:
feat: Add Schema Visualizer to Supabase Studio

AdityaTD commented:
Hey Kamil, I made a PR to add legends for constraints but require some help to gather additional data.
https://github.com/supabase/supabase/issues/15585#issuecomment-1637056995

joshenlim commented on studio/components/interfaces/Database/Schemas/SchemaGraph.tsx:
```diff
@@ -0,0 +1,341 @@
+          </header>
+        </div>
+      ) : (
+        <div className={`w-[${NODE_WIDTH / 2}px] rounded-lg overflow-hidden`}>
```
can we try to avoid this? tailwind doesn't like dynamic class names ([ref](https://tailwindcss.com/docs/content-configuration#dynamic-class-names))

joshenlim commented on studio/components/interfaces/Database/Schemas/SchemaGraph.tsx:
```diff
@@ -0,0 +1,341 @@
+  )
+}
+
+const TablesGraph: FC<{ tables: PostgresTable[] }> = ({ tables }) => {
```
we're avoiding using `FC` as well
```suggestion
const TablesGraph = ({ tables }: { tables: PostgresTable[] }) => {
```

joshenlim commented on studio/components/interfaces/Database/Schemas/SchemaGraph.tsx:
```diff
@@ -0,0 +1,341 @@
+
+  return (
+    <>
+      <div style={{ width: '100%', height: '100%' }}>
```
can use tailwind where possible
```suggestion
      <div className="w-full h-full">
```

joshenlim commented on studio/components/interfaces/Database/Schemas/SchemaGraph.tsx:
```diff
@@ -0,0 +1,341 @@
+            hideAttribution: true,
+          }}
+        >
+          <Background id="1" gap={10} color="#262626" variant={BackgroundVariant.Lines} />
```
i might be selecting the wrong line here but on light mode the grid lines against the background are standing out too much - could we use a lighter color? (and if possible using our tailwind classes for this?)

![image](https://github.com/supabase/supabase/assets/19742402/74967fbc-e062-45c2-a719-4061bade9209)

kamilogorek replied:
Updated. Unfortunately, we cannot use tailwind here.

MildTomato replied:
we could set up some tailwind custom classes to deal with some of this if we need to.
but probably no need if we just import a css file onto the NextJs page.

joshenlim commented on studio/components/interfaces/Database/Schemas/SchemaGraph.tsx:
```diff
@@ -0,0 +1,341 @@
+  )
+}
+
+const SchemaGraph: FC<{ schema: string }> = ({ schema }) => {
```
```suggestion
const SchemaGraph = ({ schema }: { schema: string }) => {
```

joshenlim commented:
This is amazing btw! great stuff @kamilogorek 🙏

joshenlim commented on studio/components/layouts/DatabaseLayout/DatabaseMenu.utils.ts:
```diff
@@ -20,6 +20,12 @@ export const generateDatabaseMenu = (
           items: [],
         },
+        {
+          name: 'Schemas',
```
Can we also change this to be "Schema Visualizer" as well? and we shift it to be below "Tables"

kamilogorek commented:
Thanks for the review @joshenlim! Will follow-up on the changes on weekend :)

kamilogorek committed:
Apply code review feedback

kamilogorek committed:
Cap maxZoom to 1.5 for better presentation and zooming experience

kamilogorek committed:
Add floating label for truncated column names

kamilogorek commented:
Added:
- Updated styles for light mode
- Cap maxZoom to 1.5 for better presentation and zooming experience
- Add floating label for truncated column names

I now consider it a good, shippable MVP, which everyone can iterate over.

joshenlim commented:
hey @kamilogorek all looks good now :) would you mind resolving the conflict on package-lock? be sure to update your branch with what we have currently on master and that you're on node 18 🙏

kamilogorek commented:
@joshenlim done, rebased and resolved

AdityaTD commented:
Amazing work @kamilogorek.
@joshenlim I'll start working on the constraint legends as soon as this PR has merged.

joshenlim commented:
just a heads up - i noticed a tiny bug whereby hovering your cursor on the column hides the column type, not sure if thats intentional? 
![image](https://github.com/supabase/supabase/assets/19742402/0c932823-7557-4deb-81ee-28b0ddf0f165)

but we can fix this later on - i'm planning to get this in today :) Also just FYI i'm adding a feature flag to this page, so if you're developing locally on this just feel free to remove it first, then add it back in before submitting future PRs :)

joshenlim commented:
gotcha on the attribution as well @kamilogorek! Supabase is now sponsoring react-flow 😄 
![image](https://github.com/supabase/supabase/assets/19742402/52dac489-fa9f-4694-be2d-eda0c52519b5)

joshenlim commented on studio/components/interfaces/Database/Schemas/SchemaGraph.tsx:
```diff
@@ -0,0 +1,354 @@
+          fitView
+          maxZoom={1.5}
+          proOptions={{
+            /**
+             * TODO: Sponsor Reactflow before using it in production! 😇
+             * https://pro.reactflow.dev/
+             */
+            hideAttribution: false,
```
```suggestion
            /**
             * TODO: Sponsor Reactflow before using it in production! 😇
             * https://pro.reactflow.dev/
             */
            hideAttribution: true,
```

joshenlim replied:
actually no worries - i'll do this in another PR while adding the feature flag

kamilogorek commented:
> i noticed a tiny bug whereby hovering your cursor on the column hides the column type, not sure if thats intentional?

Yes, that was intentional to make more room for truncated column names, but maybe there's more elegant solution :)

Lovely, thanks for the review and for taking care of the integration!
